Expand webpushd connection SPI to enable regression testing of notifications

beidson opened the pull request:
#### 8f9f5a32d4cee9eceba7c5fe2bb6ffda454b329d
<pre>
Expand webpushd connection SPI to enable regression testing of notifications
<a href="https://rdar.apple.com/135722662">rdar://135722662</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=279447">https://bugs.webkit.org/show_bug.cgi?id=279447</a>

Reviewed by Sihui Liu.

Also includes a drive-by fix of tag processing in webpushd, which was easy to test
because of this change!

* Source/WebCore/Modules/notifications/NotificationData.h:

* Source/WebKit/UIProcess/API/APIWebPushDaemonConnection.cpp:
(API::WebPushDaemonConnection::showNotification):
(API::WebPushDaemonConnection::getNotifications):
(API::WebPushDaemonConnection::cancelNotification):
* Source/WebKit/UIProcess/API/APIWebPushDaemonConnection.h:

* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:

* Source/WebKit/UIProcess/API/Cocoa/_WKNotificationData.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKNotificationData.mm:
(-[_WKNotificationData _init]):
(-[_WKNotificationData _initWithCoreData:]):
(-[_WKNotificationData _getCoreData]):
(-[_WKNotificationData setTitle:]):
(-[_WKNotificationData title]):
(-[_WKNotificationData setDir:]):
(-[_WKNotificationData dir]):
(-[_WKNotificationData setLang:]):
(-[_WKNotificationData lang]):
(-[_WKNotificationData setBody:]):
(-[_WKNotificationData body]):
(-[_WKNotificationData setTag:]):
(-[_WKNotificationData tag]):
(-[_WKNotificationData setAlert:]):
(-[_WKNotificationData alert]):
(-[_WKNotificationData setData:]):
(-[_WKNotificationData data]):
(-[_WKNotificationData origin]):
(-[_WKNotificationData setSecurityOrigin:]):
(-[_WKNotificationData securityOrigin]):
(-[_WKNotificationData setServiceWorkerRegistrationURL:]):
(-[_WKNotificationData serviceWorkerRegistrationURL]):
(-[_WKNotificationData identifier]):
(-[_WKNotificationData setUuid:]):
(-[_WKNotificationData uuid]):
(-[_WKNotificationData userInfo]):
(-[_WKNotificationData dealloc]):
(-[_WKMutableNotificationData init]):
(-[_WKNotificationData initWithCoreData:dataStore:]): Deleted.
(-[_WKNotificationData dictionaryRepresentation]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKNotificationDataInternal.h:

* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.mm:
(-[_WKWebPushDaemonConnection showNotification:completionHandler:]):
(-[_WKWebPushDaemonConnection getNotifications:tag:completionHandler:]):
(-[_WKWebPushDaemonConnection cancelNotification:uuid:]):

* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::getNotifications):

* Tools/MiniBrowser/mac/AppDelegate.m:
(-[BrowserAppDelegate createBrowserWindowController:]): Unfortunately, because Obj-C is
  what it is, and `tag` is a common selector, do gnarly Obj-C tricks here.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(TestWebKitAPI::(WebPushD, WKWebPushDaemonConnectionPushNotifications)):
(TestWebKitAPI::(WebPushD, WKWebPushDaemonConnectionPushSubscription)): Deleted.

Canonical link: <a href="https://commits.webkit.org/283493@main">https://commits.webkit.org/283493@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a5adee9c782be7a7cfa827c0c91648cb473e9754

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66432 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45807 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19053 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70465 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/17565 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68550 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53606 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17325 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53281 "Passed tests") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/17565 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69499 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42219 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57509 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33935 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38890 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14897 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15918 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60791 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15239 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72168 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10389 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14608 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60606 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10421 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57578 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60921 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14646 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8574 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/2190 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41614 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42691 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43874 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42434 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/Modules/notifications/NotificationData.h:
```diff
@@ -59,9 +59,9 @@ struct NotificationData {
-    PAL::SessionID sourceSession;
+    WTF::UUID notificationID { WTF::UUID::createVersion4() };
+    ScriptExecutionContextIdentifier contextIdentifier { ScriptExecutionContextIdentifier::generate() };
+    PAL::SessionID sourceSession { PAL::SessionID::defaultSessionID() };
```
It seems weird to generate `ScriptExecutionContextIdentifier` for `NotificationData`. We could just use `std::optional<>` or `Markable<>` if these data members may not be set.

szewai commented on Source/WebKit/UIProcess/API/APIWebPushDaemonConnection.h:
```diff
@@ -29,6 +29,7 @@
 #include "WebPushDaemonConnection.h"
 
 namespace WebCore {
+class SecurityOriginData;
```
Unused

szewai commented on Source/WebKit/UIProcess/API/Cocoa/_WKNotificationData.h:
```diff
@@ -24,20 +24,53 @@
  */
 #import <WebKit/WKFoundation.h>
 
+@class WKSecurityOrigin;
```
Unused

szewai commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.mm:
```diff
@@ -150,6 +152,40 @@ - (void)getNextPendingPushMessage:(void (^)(_WKWebPushMessage *))completionHandl
+                return completionHandlerCopy(nsResult, nil);
+            }
+
+            return completionHandlerCopy(nil, nil);
```
This should probably return 0-sized `NSArray`

beidson replied:
In our brave Swifty future, you're probably right.

szewai commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.mm:
```diff
@@ -150,6 +152,40 @@ - (void)getNextPendingPushMessage:(void (^)(_WKWebPushMessage *))completionHandl
+
+- (void)cancelNotification:(NSURL *)securityOriginURL uuid:(NSUUID *)notificationIdentifier
+{
+    // UUID::fromNSUUID only fails if the passed in NSUUID is nil, which would be a crash-worthy misuse of this API
```
Should we take a completion handler that returns NSError in this case?

beidson replied:
Our experience is generally that t's okay to have APIs crash when you blatantly misuse them (WebKit is certainly rife with them)

szewai commented on Source/WebKit/UIProcess/API/Cocoa/_WKNotificationData.h:
```diff
@@ -24,20 +24,53 @@
+@end
+
+WK_CLASS_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA))
+@interface _WKMutableNotificationData : _WKNotificationData
```
Can we just make `_WKNotificationData` properties readwrite? What's the benefit of `_WKMutableNotificationData`?

beidson replied:
Hmmmmm will stew on this offline.

beidson committed:
Expand webpushd connection SPI to enable regression testing of notifications
rdar://135722662
https://bugs.webkit.org/show_bug.cgi?id=279447

Reviewed by Sihui Liu.

Also includes a drive-by fix of tag processing in webpushd, which was easy to test
because of this change!

* Source/WebCore/Modules/notifications/NotificationData.h:

* Source/WebKit/UIProcess/API/APIWebPushDaemonConnection.cpp:
(API::WebPushDaemonConnection::showNotification):
(API::WebPushDaemonConnection::getNotifications):
(API::WebPushDaemonConnection::cancelNotification):
* Source/WebKit/UIProcess/API/APIWebPushDaemonConnection.h:

* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:

* Source/WebKit/UIProcess/API/Cocoa/_WKNotificationData.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKNotificationData.mm:
(-[_WKNotificationData _init]):
(-[_WKNotificationData _initWithCoreData:]):
(-[_WKNotificationData _getCoreData]):
(-[_WKNotificationData setTitle:]):
(-[_WKNotificationData title]):
(-[_WKNotificationData setDir:]):
(-[_WKNotificationData dir]):
(-[_WKNotificationData setLang:]):
(-[_WKNotificationData lang]):
(-[_WKNotificationData setBody:]):
(-[_WKNotificationData body]):
(-[_WKNotificationData setTag:]):
(-[_WKNotificationData tag]):
(-[_WKNotificationData setAlert:]):
(-[_WKNotificationData alert]):
(-[_WKNotificationData setData:]):
(-[_WKNotificationData data]):
(-[_WKNotificationData origin]):
(-[_WKNotificationData setSecurityOrigin:]):
(-[_WKNotificationData securityOrigin]):
(-[_WKNotificationData setServiceWorkerRegistrationURL:]):
(-[_WKNotificationData serviceWorkerRegistrationURL]):
(-[_WKNotificationData identifier]):
(-[_WKNotificationData setUuid:]):
(-[_WKNotificationData uuid]):
(-[_WKNotificationData userInfo]):
(-[_WKNotificationData dealloc]):
(-[_WKMutableNotificationData init]):
(-[_WKNotificationData initWithCoreData:dataStore:]): Deleted.
(-[_WKNotificationData dictionaryRepresentation]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKNotificationDataInternal.h:

* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.mm:
(-[_WKWebPushDaemonConnection showNotification:completionHandler:]):
(-[_WKWebPushDaemonConnection getNotifications:tag:completionHandler:]):
(-[_WKWebPushDaemonConnection cancelNotification:uuid:]):

* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::getNotifications):

* Tools/MiniBrowser/mac/AppDelegate.m:
(-[BrowserAppDelegate createBrowserWindowController:]): Unfortunately, because Obj-C is
  what it is, and `tag` is a common selector, do gnarly Obj-C tricks here.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(TestWebKitAPI::(WebPushD, WKWebPushDaemonConnectionPushNotifications)):
(TestWebKitAPI::(WebPushD, WKWebPushDaemonConnectionPushSubscription)): Deleted.

Canonical link: https://commits.webkit.org/283493@main
